src/java/org/apache/cassandra/schema/PartitionDenylist.java

```diff
@@ -0,0 +1,406 @@
+            return new DenylistEntry();
+
+        final String readDenylist = String.format("SELECT * FROM %s.%s WHERE ks_name='%s' AND cf_name='%s' LIMIT %d", SystemDistributedKeyspace.NAME, PARTITION_DENYLIST_TABLE,
+                                                  tmd.keyspace, tmd.name, DatabaseDescriptor.maxdenyDenylistKeysPerTable() + 1);
+
+        try
+        {
+            final UntypedResultSet results = process(readDenylist, DatabaseDescriptor.denylistConsistencyLevel());
+            if (results == null || results.isEmpty())
+                return new DenylistEntry();
+
+            if (results.size() > DatabaseDescriptor.maxdenyDenylistKeysPerTable())
+                logger.error("Partition denylist for {}/{} has exceeded the maximum allowable size ({}). Remaining keys were ignored.", tmd.keyspace, tmd.name, DatabaseDescriptor.maxdenyDenylistKeysPerTable());
+
+            final Set<ByteBuffer> keys = new HashSet<>();
+            final NavigableSet<Token> tokens = new TreeSet<>();
+            for (final UntypedResultSet.Row row : results)
+            {
+                final ByteBuffer key = row.getBlob("key");
+                keys.add(key);
+                tokens.add(StorageService.instance.getTokenMetadata().partitioner.getToken(key));
+            }
+            return new DenylistEntry(ImmutableSet.copyOf(keys), ImmutableSortedSet.copyOf(tokens));
+        }
+        catch (final RequestExecutionException e)
+        {
+            logger.error("Error reading partition_denylist table for {}/{}", tmd.keyspace, tmd.name, e);
+            return null;
+        }
+    }
+
+    private Map<TableId, DenylistEntry> readAll()
+    {
+        final String readDenylist = String.format("SELECT * FROM %s.%s LIMIT %d", SystemDistributedKeyspace.NAME, PARTITION_DENYLIST_TABLE, DatabaseDescriptor.maxdenyDenylistKeysTotal() + 1);
+        try
+        {
+            final UntypedResultSet results = process(readDenylist, DatabaseDescriptor.denylistConsistencyLevel());
+            if (results == null || results.isEmpty())
+                return new HashMap<>();
+
+            if (results.size() > DatabaseDescriptor.maxdenyDenylistKeysTotal())
+                logger.error("Partition denylist has exceeded the maximum allowable total size ({}). Remaining keys were ignored.", DatabaseDescriptor.maxdenyDenylistKeysTotal());
+
+            final Map<TableId, Pair<Set<ByteBuffer>, NavigableSet<Token>>> allDenylists = new HashMap<>();
+            for (final UntypedResultSet.Row row : results)
+            {
+                final TableId tid = getTableId(row.getString("ks_name"), row.getString("cf_name"));
+                if (tid == null)
+                    continue;
+                if (!allDenylists.containsKey(tid))
+                    allDenylists.put(tid, Pair.create(new HashSet<>(), new TreeSet<>()));
+                final ByteBuffer key = row.getBlob("key");
+                allDenylists.get(tid).left.add(key);
```

Comment:
nit: 
maybe that's optimized by the compiler but I'd write it like this
```
Pair<Set<ByteBuffer>, NavigableSet<Token>> pair = allDenylists.get(tid);
pair.left.add(key);
pair.right.add(StorageService.instance.getTokenMetadata().partitioner.getToken(key));
```
to ensure we do not do unnecessary lookups.

Reply:
Honestly this is just cleaner, extra lookups or no. Combined that w/computeIfAbsent.